Sources/GRPCHTTP2Core/Client/Connection/GRPCChannel.swift

```diff
@@ -0,0 +1,912 @@
+  mutating func start() {
+    precondition(!self.running, "channel must only be started once")
+    self.running = true
+  }
+
+  mutating func setNameResolverTaskHandle(_ handle: CancellableTaskHandle) {
+    switch self.state {
+    case .notRunning(var state):
+      state.nameResolverHandle = handle
+      self.state = .notRunning(state)
+    case .running, .stopping, .stopped, ._modifying:
+      fatalError("Invalid state")
+    }
+  }
+
+  enum LoadBalancerKind {
+    case roundRobin
+
+    func matches(loadBalancer: LoadBalancer) -> Bool {
+      switch (self, loadBalancer) {
+      case (.roundRobin, .roundRobin):
+        return true
+      }
+    }
+  }
+
+  enum OnChangeLoadBalancer {
+    case runLoadBalancer(LoadBalancer, stop: LoadBalancer?)
+    case updateLoadBalancer(LoadBalancer)
+    case none
+  }
+
+  mutating func changeLoadBalancerKind(
+    to newLoadBalancerKind: LoadBalancerKind,
+    _ makeLoadBalancer: () -> LoadBalancer
+  ) -> OnChangeLoadBalancer {
+    let onChangeLoadBalancer: OnChangeLoadBalancer
+
+    switch self.state {
+    case .notRunning(let state):
+      let loadBalancer = makeLoadBalancer()
+      let state = State.Running(from: state, loadBalancer: loadBalancer)
+      self.state = .running(state)
+      onChangeLoadBalancer = .runLoadBalancer(state.current, stop: nil)
+
+    case .running(var state):
+      self.state = ._modifying
+
+      if let next = state.next {
+        if newLoadBalancerKind.matches(loadBalancer: next) {
+          onChangeLoadBalancer = .updateLoadBalancer(next)
+        } else {
+          // The 'next' didn't become ready in time. Close it and replace it with a load-balancer
+          // of the next kind.
+          let nextNext = makeLoadBalancer()
+          let previous = state.next
+          state.next = nextNext
+          state.past[next.id] = next
+          onChangeLoadBalancer = .runLoadBalancer(nextNext, stop: previous)
+        }
+      } else {
+        if newLoadBalancerKind.matches(loadBalancer: state.current) {
+          onChangeLoadBalancer = .updateLoadBalancer(state.current)
+        } else {
+          // Create the 'next' load-balancer, it'll replace 'current' when it becomes ready.
+          let next = makeLoadBalancer()
+          state.next = next
+          onChangeLoadBalancer = .runLoadBalancer(next, stop: nil)
+        }
+      }
+
+      self.state = .running(state)
+
+    case .stopping, .stopped:
+      onChangeLoadBalancer = .none
+
+    case ._modifying:
+      fatalError("Invalid state")
+    }
+
+    return onChangeLoadBalancer
+  }
+
+  struct ConnectivityStateChangeActions {
+    var close: LoadBalancer? = nil
+    var publishState: ConnectivityState? = nil
+    var resumeContinuations: ResumableContinuations? = nil
+    var finish: Bool = false
+
+    struct ResumableContinuations {
+      var continuations: [CheckedContinuation<LoadBalancer, Error>]
+      var result: Result<LoadBalancer, Error>
+    }
+  }
```

Comment:
Is there a reason (thinking allocations or other performance) why we want this to be a struct vs having an enum of actions and returning an array of these actions from `loadBalancerStateChanged`? 

I feel like returning an array of states is safer in terms of handling all of the actions (i.e., if we ever add an action here, we will forced to update callers to handle it) and also the state machine itself will enforce the order in which we should handle each action, instead of leaving that up to the caller, which may make more sense?

Reply:
An array of actions would cause an allocation each time the state changed. 

I'm not sure the ordering matters much, we either have to get it right in the state machine (when building the array of actions) or when we act on the actions as we do now. This would be a bit different if the caller didn't know what the state machine did but here the objects are so tightly coupled that enforcing the order in the state machine doesn't really add anything.

Accidentally not handling a state is a valid concern. However, if we add an action we're doing it purposefully so I'd be surprised if that got missed while writing the code or in code review. IMO it's not worth adding an allocation for it.